src/vsr/replica_test.zig

```diff
@@ -0,0 +1,920 @@
+    });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.R0).corrupt(.{ .wal_header = 20 });
+    t.replica(.R0).stop();
+    try expectEqual(t.replica(.R0).open(), .ok);
+    try expectEqual(c.request(30), 30);
+    try expectEqual(t.replica(.R0).commit(), 30);
+    try expectEqual(t.replica(.S0).commit(), 30);
+}
+
+test "Cluster: network: partition 2-1 (isolate backup, symmetric)" {
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.B2).drop_all(.__, .bidirectional);
+    try expectEqual(c.request(40), 40);
+    try expectEqual(t.replica(.A0).commit(), 40);
+    try expectEqual(t.replica(.B1).commit(), 40);
+    try expectEqual(t.replica(.B2).commit(), 20);
+}
+
+test "Cluster: network: partition 2-1 (isolate backup, asymmetric, send-only)" {
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.B2).drop_all(.__, .incoming);
+    try expectEqual(c.request(40), 40);
+    try expectEqual(t.replica(.A0).commit(), 40);
+    try expectEqual(t.replica(.B1).commit(), 40);
+    try expectEqual(t.replica(.B2).commit(), 20);
+}
+
+test "Cluster: network: partition 2-1 (isolate backup, asymmetric, receive-only)" {
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.B2).drop_all(.__, .outgoing);
+    try expectEqual(c.request(40), 40);
+    try expectEqual(t.replica(.A0).commit(), 40);
+}
+
+test "Cluster: network: partition 1-2 (isolate primary, symmetric)" {
+    // The primary cannot communicate with either backup, but the backups can communicate with one
+    // another. The backups will perform a view-change since they don't receive heartbeats.
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.A0).drop_all(.B1, .bidirectional);
+    t.replica(.A0).drop_all(.B2, .bidirectional);
+    try expectEqual(c.request(30), 30);
+}
+
+test "Cluster: network: partition 1-2 (isolate primary, asymmetric, send-only)" {
+    // The primary can send to the backups, but not receive.
+    // After a short interval of not receiving messages (specifically prepare_ok's) it will abdicate
+    // by pausing heartbeats, allowing the next replica to take over as primary.
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.A0).drop_all(.B1, .incoming);
+    t.replica(.A0).drop_all(.B2, .incoming);
```

Comment:
I was thinking that (eventually) metrics counters might be an even better fit for explicit code coverage, to get a little type safety. (e.g. if metric names are an enum). But I'm listing the log messages for now to make it easier to add later.

Reply:
I am wondering if we maybe should decouple logging in replica?

so that, rather that `log.error(“string”, args)` we do `event(.tag, args)`, and it’s event who then makes a format string for log (or directly matches on tag in test)

Reply:
:thinking: Maybe with `Event = union(enum) { tag: args, ... }`?
The `event()` could handle logging, metrics, tracing, and even [simulator hooks](https://github.com/tigerbeetledb/tigerbeetle/blob/887a01f5ccbae47544d0ddfc7d8ec06a7bf958dc/src/vsr/replica.zig#L333-L345).